app/scripts/views/settings.js

```diff
@@ -3,78 +3,223 @@
+          this._onProfileChange.bind(this));
+      }
     },
 
     context: function () {
       var account = this.getSignedInAccount();
-      var email = account.get('email');
 
       return {
-        email: email,
+        username: account.get('email'),
         showSignOut: ! account.isFromSync(),
         communicationPrefsVisible: this._areCommunicationPrefsVisible()
       };
     },
 
     events: {
-      // validateAndSubmit is used to prevent multiple concurrent submissions.
-      'click #signout': BaseView.preventDefaultThen('validateAndSubmit')
+      'click #signout': BaseView.preventDefaultThen('signOut')
+    },
+
+    _onProfileChange: function () {
+      this._showAvatar();
+    },
+
+    // When we navigate to settings from a subview
+    // close the modal, show any ephemeral messages passed to `navigate`
+    _onNavigateFromSubview: function () {
+      if ($.modal.isActive()) {
+        $.modal.close();
+      }
+      this.showEphemeralMessages();
+      this.logScreen();
+    },
+
+    showSubView: function (SubView, options) {
+      var self = this;
+      if (self._panelViews.indexOf(SubView) === -1) {
+        return;
+      }
+
+      // Destroy any previous modal view
+      if (self._currentSubView && self._currentSubView.isModal) {
+        self._currentSubView.closePanel();
+      }
+
+      return self._createSubViewIfNeeded(SubView, options)
+        .then(function (subView) {
+          if (subView) {
+            self._currentSubView = subView;
+            subView.openPanel();
+            subView.logScreen();
+            return subView;
+          }
+        });
+    },
+
+    _subviewInstanceFromClass: function (SubView) {
+      return this.subviews.filter(function (subView) {
+        if (subView instanceof SubView) {
+          return true;
+        }
+      })[0];
+    },
+
+    _isModalViewClass: function (SubView) {
+      return !! SubView.prototype.isModal;
+    },
+
+    _subViewClass: function (SubView) {
+      return SubView.prototype.className;
+    },
+
+    // Render subview if an instance doesn't already exist
+    _createSubViewIfNeeded: function (SubView) {
+      var subView = this._subviewInstanceFromClass(SubView);
+      if (subView) {
+        return p(subView);
+      }
+
+      var self = this;
+      var className = self._subViewClass(SubView);
+      var selector = '.' + className;
+
+      self.$('#subviews').append('<div class="settings-subview ' + className + '"></div>');
+
+      var view = self.router.createView(SubView, {
+        el: self.$(selector),
+        superView: self
+      });
+
+      self.trackSubview(view);
+
+      return renderView(view);
     },
 
     beforeRender: function () {
-      if (this.relier.get('setting') === 'avatar') {
-        this.navigate('/settings/avatar/change');
-        return false;
+      var self = this;
+      if (self.relier.get('setting') === 'avatar') {
+        self.relier.set('setting', null);
```

Comment:
This code exists on master-- we use it when linking from the browser to the
avatar change screen.
On Aug 4, 2015 3:36 AM, "Shane Tomlinson" <notifications@github.com> wrote:

> In app/scripts/views/settings.js
> <https://github.com/mozilla/fxa-content-server/pull/2748#discussion_r36175366>
> :
>
> >      },
> >
> >      beforeRender: function () {
> > -      if (this.relier.get('setting') === 'avatar') {
> > -        this.navigate('/settings/avatar/change');
> > -        return false;
> > +      var self = this;
> > +      if (self.relier.get('setting') === 'avatar') {
> > +        self.relier.set('setting', null);
>
> What does this do?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/fxa-content-server/pull/2748/files#r36175366>.
>